### ISSUE TITLE
Bug: space-before-function-paren - to greedy match

### ISSUE THREAD

**Issue by kisfairmed** [typescript-eslint/typescript-eslint#5374]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=4.7.2&sourceType=module&code=GYVwdgxgLglg9mABMOcA8K6IKYA8rZgAmAzgBRggA2VAPpQLYBG2ATgJQB8Z7iA3gChEiVtighWSAORSBAXyA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKWQBwEMBjRAWgCNEAzAe2jKtgDsiAXASxqdMPqfUiJo0OpHBgAviAlA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA

### Repro Code

```typescript
function foo<foo extends(null|number)>() {
  return ''
}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  "rules": {
    "space-before-function-paren":"error"
  }
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    // ...
  }
}
```

### Expected Result

Error should not trigger

### Actual Result

Error: "Missing space before function parentheses" ("space-before-function-parent" Rule)

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `8.15.0` |
| `TypeScript`                           | `4.7.4` |

**Comment by JoshuaKGoldberg** [comment]
> `"space-before-function-paren":"error"`

This config is using the built-in ESLint `space-before-function-paren`. You need to use the [`@typescript-eslint/space-before-function-paren`](https://typescript-eslint.io/rules/space-before-function-paren) rule instead. It's a fixed version of the core ESLint that adds support for exactly what's in this issue.

https://typescript-eslint.io/play/#ts=4.7.2&sourceType=module&code=GYVwdgxgLglg9mABMOcCMAeFdEFMAeUuYAJgM6IAUYIANrYgD6I0C2ARrgE4CUAfJR6IA3gChEiLrigguSAOTzRAX1GjQkWAmSoATFlR5CxclRr0mLEB278qQsRKky5iRSqA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6ZYgQzMVoCNEAMwD20XoNhMy+SsKa0uYpukiJo0UZHBgAviB1A&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA
